Completed second homework

juhiaugustine opened the pull request:
**Exercise 01 - Coins**

<img width="1209" alt="Screen Shot 2022-10-19 at 4 30 17 PM" src="https://user-images.githubusercontent.com/114516950/196824427-1b4e58a1-d763-47c7-b08e-32631a2ad1b9.png">

**Exercise 02 - Reverse**

<img width="1512" alt="Screen Shot 2022-10-19 at 4 31 21 PM" src="https://user-images.githubusercontent.com/114516950/196824444-bf528ef5-474a-4824-a567-264edaaf7301.png">
<img width="1511" alt="Screen Shot 2022-10-19 at 4 31 35 PM" src="https://user-images.githubusercontent.com/114516950/196824457-ce901351-81cd-461a-88f6-d5ffbc94adcc.png">

**Exercise 03 - Form**

<img width="1512" alt="Screen Shot 2022-10-19 at 4 36 35 PM" src="https://user-images.githubusercontent.com/114516950/196824515-3d89a69a-0ed1-406b-adf4-3b675f174f6d.png">
<img width="1512" alt="Screen Shot 2022-10-19 at 4 34 21 PM" src="https://user-images.githubusercontent.com/114516950/196824527-3d4e3b36-a8f3-484d-b4f8-1f924d7de598.png">
<img width="1512" alt="Screen Shot 2022-10-19 at 4 34 33 PM" src="https://user-images.githubusercontent.com/114516950/196824534-3c25c52b-3607-4230-ad3f-afae02f16e6d.png">

**Exercise 04 - API**

<img width="1512" alt="Screen Shot 2022-10-19 at 4 35 26 PM" src="https://user-images.githubusercontent.com/114516950/196824549-9997590a-9992-49ca-9f74-3ade2fb234df.png">
<img width="1512" alt="Screen Shot 2022-10-19 at 4 38 53 PM" src="https://user-images.githubusercontent.com/114516950/196824555-7105db65-3cc1-404d-b457-fc8f00c88a8f.png">

gleason9113 reviewed (commented):
Clean, beautiful code.  I see nothing to improve on!
Approved!

CzerPDX commented on hw2/01-coins.js:
```diff
@@ -1,7 +1,52 @@
+  dollarString = dollars == 0 ? "" 
+                : dollars == 1 ? dollars.toString() + " dollar, "
+                : dollars.toString() + " dollars, ";
+
```
I love how efficient this code is! I think your method of dividing the amount by the value and then doing a floor is so clever. I wrote mine almost like how you would count them in real life. I think yours is very elegant. The only thing I could suggest is encapsulating some of this into functions.

Code produces expected output!

CzerPDX commented on hw2/02-reverse.js:
```diff
@@ -1,3 +1,21 @@
 
 // Add your code here
+function reverseNumber() {
+    let inputValue = parseInt(document.getElementById("input").value, 10);
```
Code is again very nicely written and formatted. After running it produce expected output with one exception: I am unable to reverse 8-digit numbers with leading 0's.

For example: entering `00001111` 
produces `Error: Please input an 8-digit number`
instead of `11110000`.

However, other than that output difference this code is easily readable and efficient.

CzerPDX commented on hw2/03-form.js:
```diff
@@ -1,3 +1,24 @@
 // Add your code here
+
+function submitForm(){
+    console.log("========== Form Submission ===========");
```
Nice use of bootstrap form styling. I made some of the same choices you did for those.

When running this code I noticed a few validation issues, though
- There is no validation to stops the user from submitting an empty form. The name and email fields are required, so some checking should be added to make sure they've been entered by the user. 
- There is no validation to check whether the user entered a valid email. I was able to enter random text as the email when I tested
- Additionally, I believe the requirements mentioned error messages when incorrect values are put in. So, I think it should give you feedback when you don't pass the validation. 
    - [This documentation article](https://getbootstrap.com/docs/5.0/forms/validation/) was really helpful for me and I was able to even directly borrow the example function for disabling the submit until all fields were filled out.

Lastly, another student said we were supposed to use `console.group()` for the output, but I'm not 100% sure if that's correct. I ended up implementing it anyway, as it seemed to be a useful way to group things and it matched the professor's output in the example, so that that as you will!

CzerPDX commented on hw2/04-api.js:
```diff
@@ -1,7 +1,51 @@
-// Add your code here
+/* 
+    Loops through the JSON and formats each element as a string
+    storing the country name and population. 
```
Produces expected output perfectly. Code is well commented and easy to read. I can see no suggestions at all for change here looks great!

juhiaugustine replied:
Thanks for your inputs it was very informative. I have addressed all your comments in my latest [commit](https://github.com/juhiaugustine/fullstack-homework/pull/3/commits/863e40b3b1b02ee6444008d3ea4b426a59b3954f). Please have a look. @CzerPDX

CzerPDX replied:
Your changes look beautiful! Approved.

CzerPDX reviewed (commented):
A few comments on your files.

rk-pdx reviewed (changes requested):
`01-coins.js`

Code is clean and readable; code is structured in a logical manner; code produces the correct output. I especially like the concise way in which you derived `resultString`.

**[changes requested]** The output needs to be in the form "[original_input] => ...", whereas your code lists the coin values. 


`02-reverse.html` and `02-reverse.js`

Code is clean and readable; code produces the correct output. Good job covering all the edge cases. `02-reverse.js` can be very slightly improved stylistically (simple whitespace between variable declarations and rest of code, etc.). Overall, the code is great in terms of design, organization and readability.

`03-form.html` and `03-form.js`

Once again, the code is excellent in terms of organization, logic and readability. One small thing that I would consider changing is to provide clearer feedback when the user fails to enter both the name and the email fields. Currently, if a user leaves both fields blank and tries to submit the form, an alert pops up that instructs them to enter their name. It would be slightly better if the user got feedback that instructed them that both required fields were missing.

`04-api.html` and `04-api.js`

Nothing to comment - everything looks great.

CzerPDX reviewed (approved):
Approved after changes!

Saylee119 reviewed (approved):
Code looks good.

juhiaugustine commented:
> `01-coins.js`
> 
> Code is clean and readable; code is structured in a logical manner; code produces the correct output. I especially like the concise way in which you derived `resultString`.
> 
> **[changes requested]** The output needs to be in the form "[original_input] => ...", whereas your code lists the coin values.
> 
> `02-reverse.html` and `02-reverse.js`
> 
> Code is clean and readable; code produces the correct output. Good job covering all the edge cases. `02-reverse.js` can be very slightly improved stylistically (simple whitespace between variable declarations and rest of code, etc.). Overall, the code is great in terms of design, organization and readability.
> 
> `03-form.html` and `03-form.js`
> 
> Once again, the code is excellent in terms of organization, logic and readability. One small thing that I would consider changing is to provide clearer feedback when the user fails to enter both the name and the email fields. Currently, if a user leaves both fields blank and tries to submit the form, an alert pops up that instructs them to enter their name. It would be slightly better if the user got feedback that instructed them that both required fields were missing.
> 
> `04-api.html` and `04-api.js`
> 
> Nothing to comment - everything looks great.

Hello @rk-pdx , Thanks for your detailed PR review. I have addressed your comments in my latest code. Please take a look at these [changes](https://github.com/juhiaugustine/fullstack-homework/pull/3/files/863e40b3b1b02ee6444008d3ea4b426a59b3954f..67cbe4b8ac14dd14e5470f70e5b058283137ac6f)

JeremyIV commented:
Very nice job, Juhi!
